File: 121. 买卖股票的最佳时机.cpp



#if 0

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0;
        if(prices.size() == 0)
            return -1;


        int minLeft;
        minLeft = prices[0];


        for(int i=1;i<prices.size();++i)
        {
            minLeft = min(minLeft,prices[i]);
            res = max(res,prices[i] - minLeft);


        }


        return res;

    }
};


int main() {
    Solution s;
    vector<int> v1{1,4,2};
    vector<int> v2{7,6,4,3,1};
        cout << s.maxProfit(v1) << endl;
        cout << s.maxProfit(v2) << endl;



    return 0;
}


#endif

File: leetcode100/121. 买卖股票的最佳时机.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
@Project ：LeetCodePthonVersion 
@File ：121. 买卖股票的最佳时机.py
@Author ：HuntingGame
@Date ：2023-02-05 11:27 
C'est la vie!!! enjoy ur day :D
'''
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans= -prices[0]
        minval = prices[0]
        for i in range(1,len(prices)):
            ans = max(ans,prices[i] - minval)
            minval = min(minval,prices[i])


        return ans